tensorboard/uploader/uploader.py

```diff
@@ -445,19 +485,146 @@ def _create_point(self, tag_proto, event, value):
         return point
 
 
+class _BlobRequestSender(object):
+    """Uploader for blob-type event data.
+
+    Unlike the other types, this class does not accumulate events in batches;
+    every blob is sent individually and immediately.  Nonetheless we retain
+    the `add_event()`/`flush()` structure for symmetry.
+
+    This class is not threadsafe. Use external synchronization if calling its
+    methods concurrently.
+    """
+
+    def __init__(self, experiment_id, api, rpc_rate_limiter):
+        if experiment_id is None:
+            raise ValueError("experiment_id cannot be None")
+        self._experiment_id = experiment_id
+        self._api = api
+        self._rpc_rate_limiter = rpc_rate_limiter
+
+        self._new_request()
+
+    def _new_request(self):
+        """Declares the previous event complete."""
+        self._run_name = None
+        self._event = None
+        self._value = None
+        self._metadata = None
+
+    def add_event(
+        self, run_name, event, value, metadata,
+    ):
+        """Attempts to add the given event to the current request.
+
+        If the event cannot be added to the current request because the byte
+        budget is exhausted, the request is flushed, and the event is added
+        to the next request.
+        """
+        if self._value:
+            raise RuntimeError("Tried to send blob while another is pending")
+        self._run_name = run_name
+        self._event = event  # provides step and possibly plugin_name
+        self._value = value
+        self._metadata = metadata
+        self.flush()
+
+    def flush(self):
+        """Sends the current blob sequence fully, and clears it to make way for the next.
+        """
+        if self._value:
+            blob_sequence_id = self._get_or_create_blob_sequence()
+            print("Flushing to blob sequence id: ", blob_sequence_id)
+
+            # TODO(soergel): should we really unpack the tensor here, or ship
+            # it wholesale and unpack server side, or something else?
+            blobs = tensor_util.make_ndarray(self._value.tensor)
+            for seq_index, blob in np.ndenumerate(blobs):
+                # Note the _send_blob() stream is internally rate-limited.
+                self._rpc_rate_limiter.tick()
+                self._send_blob(blob_sequence_id, seq_index[0], blob)
+
+        self._new_request()
+
+    def _get_or_create_blob_sequence(self):
+        request = write_service_pb2.GetOrCreateBlobSequenceRequest(
+            experiment_id=self._experiment_id,
+            run=self._run_name,
+            tag=self._value.tag,
+            step=self._event.step,
+            final_sequence_length=1,
```

Comment:
Final sequence length should not always be 1, right? cf. `_tensor_size`
in `event_processing/data_provider.py`.

Reply:
Fixed